cleo_sdk/CLEO_Utils.h

```diff
@@ -112,34 +114,84 @@ namespace CLEO
+            return _strnicmp(str.data(), prefix.data(), prefix.length()) == 0;
+        }
+    }
+
     static std::string ScriptInfoStr(CLEO::CRunningScript* thread)
     {
         std::string info(1024, '\0');
         CLEO_GetScriptInfoStr(thread, true, info.data(), info.length());
         return std::move(info);
     }
 
-    // does file path points inside game directories? (game root or user files)
+    // Normalize filepath, collapse all parent directory references. Input should be absolute path without expandable %variables%
+    static void NormalizeFilepath(std::string& path, bool normalizeCase = true)
+    {
+        if (path.empty()) return;
+
+        std::replace(path.begin(), path.end(), '/', '\\');
+        if (normalizeCase) std::transform(path.begin(), path.end(), path.begin(), [](unsigned char c) { return tolower(c); }); // to lower case
+
+        // collapse references to parent directory
+        const auto ParentRef = "\\..\\";
+        const auto ParentRefLen = 4;
+
+        size_t refPos = path.find(ParentRef);
+        while (refPos != std::string::npos && refPos > 0)
+        {
+            size_t parentPos = path.rfind('\\', refPos - 1); // find start of parent name
+
+            if (parentPos == std::string::npos)
+                return; // parent must be root of the path then. We want to keep absolute path, let it be as is (even if "C:\..\" makes no sense)
+
+            path.replace(parentPos, (refPos - parentPos) + ParentRefLen - 1, ""); // remove parent and parent reference
+
+            refPos = path.find(ParentRef); // find next
+        }
+
+        while(path.back() == '\\') path.pop_back(); // remove trailing path separator(s)
+    }
+
+    // does normalized file path points inside game directories? (game root or user files)
     static bool IsFilepathSafe(CLEO::CRunningScript* thread, const char* path)
     {
-        auto IsSubpath = [](std::filesystem::path path, std::filesystem::path base)
+        if (strchr(path, '%') != nullptr)
         {
-            auto relative = std::filesystem::relative(path, base);
-            return !relative.empty() && *relative.begin() != "..";
-        };
+            return false; // do not allow paths containing expandable variables
+        }
 
-        auto fsPath = std::filesystem::path(path);
-        if (!fsPath.is_absolute())
+        std::string absolute;
+        if (!std::filesystem::path(path).is_absolute())
         {
-            fsPath = CLEO_GetScriptWorkDir(thread) / fsPath;
+            absolute = CLEO_GetScriptWorkDir(thread);
+            absolute += '\\';
+            absolute += path;
+            NormalizeFilepath(absolute, false);
+            path = absolute.c_str();
         }
 
-        if (IsSubpath(fsPath, Gta_Root_Dir_Path) || IsSubpath(fsPath, Gta_User_Dir_Path))
+        // check prefix
+        if (!StringStartsWith(path, std::string_view(Gta_Root_Dir_Path, strlen(Gta_Root_Dir_Path) - 1), false) && // without ending separator
```

Comment:
read Gta_Root_Dir_Path once on init, strip trailing / and store into a separate static variable. No need to normalize it every time

Reply:
This variable is not initialized by game on the startup.

Reply:
how can you guarantee that this variable is initialized by the time you access it?

Reply:
Rweritten